[WIP] sql: make DROP compute the list of targets at execute time

knz opened the pull request:
First commits from #30544.
Found while investigating #30409.

knz committed:
sql: fix EXPLAIN ALTER DATABASE RENAME

Prior to this patch EXPLAINing a database rename would actually
perform the rename (which is incorrect). This patch fixes it.

Release note (bug fix): EXPLAIN ALTER DATABASE .. RENAME .. does not
rename the database any more.

knz committed:
sql: make DROP compute the list of targets at execute time

... instead of prepare time. For example, when dropping a table with
dependencies, each execute of the DROP should drop the dependencies at
the time of execute, not the original dependencies at the time of
prepare.

Release note (bug fix): DROP now computes the list of target objects
when it executes, not when it is prepared.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/30545)
<!-- Reviewable:end -->

knz committed:
sql: make DROP DATABASE compute the list of targets at execute time

... instead of prepare time. For example, when dropping a database,
each execute of the DROP DATABASE should drop the database with its contents at
the time of execute, not the original contents at the time of prepare.

Release note (bug fix): DROP DATABASE now computes the list of target
objects when it executes, not when it is prepared.

knz committed:
sql: fix EXPLAIN ALTER TABLE RENAME

Prior to this patch EXPLAINing a table rename would actually
perform the rename (which is incorrect). This patch fixes it.

Release note (bug fix): EXPLAIN ALTER TABLE .. RENAME .. does not
rename the database any more.

knz committed:
sql: fix EXPLAIN ALTER TABLE RENAME COLUMN

Prior to this patch EXPLAINing a column rename would actually
perform the rename (which is incorrect). This patch fixes it.

Release note (bug fix): EXPLAIN ALTER TABLE .. RENAME COLUMN .. does
not rename the column any more.

knz committed:
sql: fix EXPLAIN ALTER INDEX RENAME

Prior to this patch EXPLAINing an index rename would actually
perform the rename (which is incorrect). This patch fixes it.

Release note (bug fix): EXPLAIN ALTER INDEX .. RENAME .. does
not rename the index any more.

knz commented:
This is incomplete -- need to add the cases for indexes views and sequences.
